File: src/core/config-loader.ts
/**
 * Configuration Loader for Qin
 * 使用 c12 加载配置，支持多种格式和继承
 * 支持自动检测项目类型和插件
 */

import { loadConfig } from "c12";
import { join } from "path";
import { existsSync, readdirSync, statSync } from "fs";
import type { QinConfig, ParsedEntry, ValidationResult } from "../types";
import { PluginDetector, type DetectionResult } from "./plugin-detector";

export class ConfigLoader {
  private cwd: string;

  constructor(cwd?: string) {
    this.cwd = cwd || process.cwd();
  }

  /**
   * Load configuration using c12
   * 支持：qin.config.ts, qin.config.js, qin.config.json, .qinrc
   * 支持：extends 继承配置
   * 支持：自动检测项目类型
   */
  async load(): Promise<QinConfig & { _detected?: DetectionResult }> {
    try {
      const { config } = await loadConfig<QinConfig>({
        name: "qin",
        cwd: this.cwd,
        defaults: {},
      });

      // 自动检测项目类型
      const detector = new PluginDetector(this.cwd);
      const detection = await detector.detect();

      // 如果没有配置文件但检测到了项目内容，使用检测结果
      const hasConfig = config && Object.keys(config).length > 0;
      const hasDetection = detection.languages.length > 0 || detection.features.length > 0;

      if (!hasConfig && !hasDetection) {
        throw new Error(
          "No project detected. Create src/Main.java or run 'qin init' to start a new project."
        );
      }

      // 零配置模式：没有配置文件但检测到了项目
      if (!hasConfig && hasDetection) {
        // 完全依赖自动检测，这是零配置的核心！
      }

      // 合并配置：用户配置优先，检测结果作为默认值
      const mergedConfig: QinConfig = {
        ...config,
        // 自动设置入口（用户配置优先）
        entry: config?.entry || detection.entry || this.findEntryOptional(),
        // 自动设置前端配置
        client: config?.client || (detection.clientDir ? { root: detection.clientDir } : undefined),
      };

      // 如果有 Java 文件但没有入口，尝试查找
      if (detection.languages.includes("java") && !mergedConfig.entry) {
        mergedConfig.entry = detection.entry;
      }

      // 验证配置（对于纯前端项目，entry 可以为空）
      const validation = this.validateWithDetection(mergedConfig, detection);
      if (!validation.valid) {
        throw new Error(`Invalid configuration: ${validation.errors.join(", ")}`);
      }

      const finalConfig = this.applyDefaults(mergedConfig);
      
      // 附加检测结果供 CLI 使用
      return {
        ...finalConfig,
        _detected: detection,
      };
    } catch (error) {
      if (error instanceof Error && error.message.includes("not found")) {
        throw error;
      }
      throw error;
    }
  }

  /**
   * 可选的入口查找（不抛出错误）
   */
  private findEntryOptional(): string | undefined {
    try {
      return this.findEntry();
    } catch {
      return undefined;
    }
  }

  /**
   * 带检测结果的验证
   */
  private validateWithDetection(config: QinConfig, detection: DetectionResult): ValidationResult {
    const errors: string[] = [];

    if (!config) {
      errors.push("Configuration is empty");
      return { valid: false, errors };
    }

    // 如果检测到 Java 但没有入口，报错
    if (detection.languages.includes("java") && !config.entry) {
      errors.push("Detected Java files but no entry point found. Please specify 'entry' in qin.config.ts");
    }

    // 如果指定了入口，验证格式
    if (config.entry && !config.entry.endsWith(".java")) {
      errors.push("'entry' must be a .java file");
    }

    if (config.dependencies && typeof config.dependencies !== "object") {
      errors.push("'dependencies' must be an object");
    }

    // 纯前端项目（没有 Java）是允许的
    if (!detection.languages.includes("java") && detection.features.includes("frontend")) {
      // 纯前端项目，不需要 entry
      return { valid: true, errors: [] };
    }

    return {
      valid: errors.length === 0,
      errors,
    };
  }

  /**
   * Auto-detect entry file
   * Search order:
   * 1. src/Main.java
   * 2. src/server/Main.java
   * 3. src/{subdir}/Main.java (any subdirectory)
   */
  findEntry(): string {
    const candidates = ["src/Main.java", "src/server/Main.java"];

    // Check fixed paths first
    for (const candidate of candidates) {
      if (existsSync(join(this.cwd, candidate))) {
        return candidate;
      }
    }

    // Search src/*/Main.java
    const srcDir = join(this.cwd, "src");
    if (existsSync(srcDir)) {
      try {
        const entries = readdirSync(srcDir);
        for (const entry of entries) {
          const subDir = join(srcDir, entry);
          if (statSync(subDir).isDirectory()) {
            const mainJava = join(subDir, "Main.java");
            if (existsSync(mainJava)) {
              return `src/${entry}/Main.java`;
            }
          }
        }
      } catch {
        // Ignore errors
      }
    }

    throw new Error(
      "No entry file found. Please specify 'entry' in qin.config.ts or create src/Main.java"
    );
  }

  /**
   * Validate configuration
   */
  validate(config: QinConfig): ValidationResult {
    const errors: string[] = [];

    if (!config) {
      errors.push("Configuration is empty");
      return { valid: false, errors };
    }

    if (!config.entry) {
      errors.push("'entry' field is required");
    } else if (!config.entry.endsWith(".java")) {
      errors.push("'entry' must be a .java file");
    }

    if (config.dependencies && typeof config.dependencies !== "object") {
      errors.push("'dependencies' must be an object");
    }

    return {
      valid: errors.length === 0,
      errors,
    };
  }

  /**
   * Apply default values to configuration
   */
  private applyDefaults(config: QinConfig): QinConfig {
    const entry = config.entry;
    // jarName: 优先使用配置值，否则根据 name 生成，最后使用 app.jar
    const jarName =
      config.output?.jarName || (config.name ? `${config.name}.jar` : "app.jar");

    const result: QinConfig = {
      ...config,
      output: {
        dir: config.output?.dir || "dist",
        jarName,
      },
    };

    // 只有在有 Java 入口时才设置 Java 相关配置
    if (entry) {
      result.entry = entry;
      result.java = {
        version: config.java?.version || "17",
        sourceDir: config.java?.sourceDir || this.parseEntry(entry).srcDir,
      };
    }

    return result;
  }

  /**
   * Parse entry path to extract source directory and class name
   */
  parseEntry(entry?: string): ParsedEntry {
    if (!entry) {
      return { srcDir: "src", className: "Main", filePath: "src/Main.java" };
    }

    // Normalize path separators
    const normalized = entry.replace(/\\/g, "/");

    // Extract file name and directory
    const lastSlash = normalized.lastIndexOf("/");
    const fileName =
      lastSlash >= 0 ? normalized.substring(lastSlash + 1) : normalized;
    const srcDir = lastSlash >= 0 ? normalized.substring(0, lastSlash) : ".";

    // Remove .java extension to get simple class name
    const simpleClassName = fileName.replace(/\.java$/, "");

    // Try to read package declaration from file
    let className = simpleClassName;
    try {
      const filePath = join(this.cwd, entry);
      const content = require("fs").readFileSync(filePath, "utf-8");
      const packageMatch = content.match(/^\s*package\s+([\w.]+)\s*;/m);
      if (packageMatch) {
        className = `${packageMatch[1]}.${simpleClassName}`;
      }
    } catch {
      // If file can't be read, use simple class name
    }

    return {
      srcDir,
      className,
      filePath: entry,
    };
  }
}

/**
 * Standalone function for parsing entry path
 */
export function parseEntryPath(entry: string): ParsedEntry {
  const loader = new ConfigLoader();
  return loader.parseEntry(entry);
}

File: src/java-rewrite/com/qin/core/ConfigLoader.java
package com.qin.core;

import com.qin.constants.QinConstants;
import com.qin.types.*;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.regex.*;

/**
 * Configuration Loader for Qin (Java 25)
 * 加载配置文件，QinConfig 现在是不可变 Record
 */
public class ConfigLoader {
    private final String cwd;
    private final Gson gson;

    public ConfigLoader() {
        this(QinConstants.getCwd());
    }

    public ConfigLoader(String cwd) {
        this.cwd = cwd;
        this.gson = new GsonBuilder().setPrettyPrinting().create();
    }

    /**
     * Load configuration from qin.config.json
     */
    public QinConfig load() throws IOException {
        // 尝试加载 JSON 配置
        Path jsonConfig = Paths.get(cwd, QinConstants.CONFIG_FILE);
        if (Files.exists(jsonConfig)) {
            String content = Files.readString(jsonConfig);
            QinConfig config = gson.fromJson(content, QinConfig.class);
            return applyDefaults(config);
        }

        // 自动检测项目类型
        PluginDetector detector = new PluginDetector(cwd);
        DetectionResult detection = detector.detect();

        if (detection.getLanguages().isEmpty() && detection.getFeatures().isEmpty()) {
            throw new IOException(
                    "No project detected. Create src/Main.java or create qin.config.json");
        }

        // 零配置模式 - 使用简化构造器
        return applyDefaults(new QinConfig(
                detectProjectName(),
                "1.0.0"));
    }

    /**
     * Auto-detect project name from directory
     */
    private String detectProjectName() {
        return Paths.get(cwd).getFileName().toString();
    }

    /**
     * Auto-detect entry file
     */
    public String findEntry() {
        String[] candidates = {
                "src/main/java/com/subhuti/Main.java",
                "src/Main.java",
                "src/server/Main.java",
                "src/App.java",
                "src/Application.java"
        };

        for (String candidate : candidates) {
            if (Files.exists(Paths.get(cwd, candidate))) {
                return candidate;
            }
        }

        // Search src/*/Main.java
        Path srcDir = Paths.get(cwd, "src");
        if (Files.exists(srcDir) && Files.isDirectory(srcDir)) {
            try (DirectoryStream<Path> stream = Files.newDirectoryStream(srcDir)) {
                for (Path entry : stream) {
                    if (Files.isDirectory(entry)) {
                        Path mainJava = entry.resolve("Main.java");
                        if (Files.exists(mainJava)) {
                            return "src/" + entry.getFileName() + "/Main.java";
                        }
                    }
                }
            } catch (IOException e) {
                // Ignore
            }
        }

        return null;
    }

    /**
     * Validate configuration
     */
    public ValidationResult validate(QinConfig config) {
        List<String> errors = new ArrayList<>();

        if (config == null) {
            errors.add("Configuration is empty");
            return ValidationResult.failure(errors);
        }

        if (config.entry() != null && !config.entry().endsWith(".java")) {
            errors.add("'entry' must be a .java file");
        }

        return errors.isEmpty() ? ValidationResult.success() : ValidationResult.failure(errors);
    }

    /**
     * Apply default values to configuration
     * 因为 Record 不可变，所以返回新的 QinConfig 实例
     */
    private QinConfig applyDefaults(QinConfig config) {
        // 如果已经有完整配置，直接返回
        if (config.entry() != null && config.output() != null && config.java() != null) {
            return config;
        }

        // 构建带默认值的新 config
        String entry = config.entry() != null ? config.entry() : findEntry();
        OutputConfig output = config.output() != null ? config.output() : new OutputConfig();
        JavaConfig java = config.java() != null ? config.java() : new JavaConfig("25");

        // 创建新的不可变配置
        return new QinConfig(
                config.name(),
                config.version(),
                config.description(),
                config.scope(),
                config.port(),
                config.localRep(),
                config.client(),
                config.plugins(),
                entry,
                config.dependencies(),
                config.devDependencies(),
                config.packages(),
                output,
                java,
                config.graalvm(),
                config.frontend(),
                config.scripts(),
                config.repositories());
    }

    /**
     * Parse entry path to extract source directory and class name
     */
    public ParsedEntry parseEntry(String entry) {
        if (entry == null) {
            return new ParsedEntry(QinConstants.DEFAULT_SOURCE_DIR, QinConstants.DEFAULT_MAIN_CLASS,
                    QinConstants.DEFAULT_SOURCE_DIR + "/Main.java");
        }

        String normalized = entry.replace("\\", "/");
        int lastSlash = normalized.lastIndexOf("/");

        String fileName = lastSlash >= 0 ? normalized.substring(lastSlash + 1) : normalized;
        String srcDir = lastSlash >= 0 ? normalized.substring(0, lastSlash) : ".";
        String simpleClassName = fileName.replace(".java", "");

        // Try to read package declaration
        String className = simpleClassName;
        try {
            Path filePath = Paths.get(cwd, entry);
            if (Files.exists(filePath)) {
                String content = Files.readString(filePath);
                Pattern pattern = Pattern.compile("^\\s*package\\s+([\\w.]+)\\s*;", Pattern.MULTILINE);
                Matcher matcher = pattern.matcher(content);
                if (matcher.find()) {
                    className = matcher.group(1) + "." + simpleClassName;
                }
            }
        } catch (IOException e) {
            // Use simple class name
        }

        return new ParsedEntry(srcDir, className, entry);
    }
}
